nano/node/vote_router.cpp

```diff
@@ -0,0 +1,190 @@
+#include <nano/lib/thread_roles.hpp>
+#include <nano/lib/utility.hpp>
+#include <nano/node/active_transactions.hpp>
+#include <nano/node/election.hpp>
+#include <nano/node/vote_cache.hpp>
+#include <nano/node/vote_router.hpp>
+
+#include <chrono>
+
+using namespace std::chrono_literals;
+
+nano::stat::detail nano::to_stat_detail (nano::vote_code code)
+{
+	auto value = magic_enum::enum_cast<nano::stat::detail> (magic_enum::enum_name (code));
+	debug_assert (value);
+	return value.value_or (nano::stat::detail{});
+}
+
+nano::stat::detail nano::to_stat_detail (nano::vote_source source)
+{
+	auto value = magic_enum::enum_cast<nano::stat::detail> (magic_enum::enum_name (source));
+	debug_assert (value);
+	return value.value_or (nano::stat::detail{});
+}
+
+nano::vote_router::vote_router (nano::vote_cache & cache, nano::recently_confirmed_cache & recently_confirmed) :
+	cache{ cache },
+	recently_confirmed{ recently_confirmed }
+{
+}
+
+nano::vote_router::~vote_router ()
+{
+	// Thread must be stopped before destruction
+	debug_assert (!thread.joinable ());
+}
+
+void nano::vote_router::insert (nano::block_hash const & hash, std::weak_ptr<nano::election> election)
+{
+	std::unique_lock lock{ mutex };
+	elections.insert_or_assign (hash, election);
+}
+
+void nano::vote_router::erase (nano::election const & election)
+{
+	std::unique_lock lock{ mutex };
+	for (auto const & [hash, _] : election.blocks ())
+	{
+		elections.erase (hash);
```

Comment:
Election can only hold a limited number of blocks and can drop blocks during its lifetime, so this erase function is not robust.